lib/Optimizer/Transforms/LinearCtrlRelations.cpp

```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_LINEARCTRLRELATIONS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "linear-ctrl-relations"
+
+using namespace mlir;
+
+namespace {
+class ThreadControl : public OpRewritePattern<quake::ToControlOp> {
```

Comment:
Will this ever be running on two different `quake::ToControlOp`s in parallel? If so, how are race conditions prevented if the multiple controls coming from different `quake::ToControlOp`s are going into the same downstream operation?

Reply:
It should be locking on the `FuncOp`. So we can run on multiple `ToControlOp`s at the same time, but they ought to be in different functions.